lib/ansible/galaxy/token.py

```diff
@@ -35,6 +48,93 @@
 
 display = Display()
 
+AH_TOKEN_EXPIRES_N_MIN_EARLY = 1
+AH_AUTH_URL = "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token"
+AH_JWKS_URI = "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/certs"
+
+# https://www.rfc-editor.org/rfc/rfc7518#section-3.1
+REQUIRED_JWT_HEADER_KEYS = set(['alg', 'typ', 'kid'])
+REQUIRED_JWT_HEADER_VALUES = {
+    'alg': ['RS256'],
+    'typ': ['JWT'],
+}
+
+PublicNumbers = namedtuple("PublicNumbers", ["e", "n"])
+
+
+def validate_jwt_format(data):
+    if len(data.split('.')) != 3:
+        raise InvalidSignature(f"unsupported format: {data}")
+
+    if any(whitespace in data.rsplit('.')[0] for whitespace in string.whitespace):
```

Comment:
`data.rsplit('.')[0]` is evaluated on each iteration and is probably not the cheapest operation here... Maybe reverse the order?

```suggestion
    if any(
        possible_whitespace in string.whitespace
        for possible_whitespace in data.rsplit('.')[0]
    ):
```

Also, can we check against the allowlist rather than a blocklist? Is there a spec for this? Is it mostly ASCII?

Reply:
Fixed.

Yeah, it's mostly ascii https://datatracker.ietf.org/doc/html/rfc4648#section-5, but we already validate it by decoding it,  so I don't think it's necessary to hardcode a valid character set. Whitespace is invalid (validation steps: https://www.rfc-editor.org/rfc/rfc7515#section-5.2), but base64.urlsafe_b64decode strips it out, so I added this check first.